### PR TITLE
`Development`: Move lecture exercise unit features into the lecture micro service 

### PR DESCRIPTION
**Pull request by MerlinMehmed**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] *This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [ ] *Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
- [ ] * not related
### Motivation and Context

This PR moves functionality from Artemis into the Lecture microservice. It moves the lecture exercise unit resource and its tests.

### Description
The PR doesn't add new features, we just want to make sure the old ones are still working in the new architecture.

### Steps for Testing

Unfortunately, it is still not possible to test this PR on a test server. It should be tested locally. 
In order to do it, you need to run:
- JHipster Registry by running `docker-compose -f src/main/docker/jhipster-registry.yml up`
- Gateway application by running `./gradlew :gateway:bootRun`
- Lecture application by running `./gradlew :lecture:bootRun --args='--spring.profiles.active=artemis,dev'`
- Artemis as usual or i.e. `./gradlew bootRun --args='--spring.profiles.active=dev,bamboo,bitbucket,artemis,scheduling' -x webapp` to skip the client code build since it's not needed

Prerequisites:
- 1 Instructor
- 1 Student

0. Open [http://localhost:8089/](http://localhost:8089/)
1. Log in to Artemis
2. Navigate to Course Management -> Lectures -> Units
5. Create a new Exercise Unit
6. Open it, make sure you can access it 
7. Check that you can access it as a student, too
8. Delete the linked exercise from "Course Management" 
9. Open again the list of lecture units and make sure the exercise is not lister there

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### CODE REVIEW

File: lecture/src/main/java/de/tum/in/www1/artemis/lecture/web/rest/ExerciseUnitResource.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        // persist lecture unit before lecture to prevent "null index column for collection" error
+        exerciseUnit.setLecture(null);
+        exerciseUnit = exerciseUnitRepository.saveAndFlush(exerciseUnit);
+        exerciseUnit.setLecture(lecture);
+        lecture.addLectureUnit(exerciseUnit);
+        Lecture updatedLecture = lectureRepository.save(lecture);
+        ExerciseUnit persistedExerciseUnit = (ExerciseUnit) updatedLecture.getLectureUnits().get(updatedLecture.getLectureUnits().size() - 1);
+
+        return ResponseEntity.created(new URI("/api/exercise-units/" + persistedExerciseUnit.getId()))
+                .headers(HeaderUtil.createEntityCreationAlert(applicationName, true, ENTITY_NAME, "")).body(persistedExerciseUnit);
```
Comment (b-fein) [inline lecture/src/main/java/de/tum/in/www1/artemis/lecture/web/rest/ExerciseUnitResource.java:85]: Same translation issue as commented on the video-units applies here as well.

File: lecture/src/main/java/de/tum/in/www1/artemis/lecture/web/rest/ExerciseUnitResource.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        if (lectureOptional.isEmpty()) {
+            throw new BadRequestAlertException("Invalid id", ENTITY_NAME, "not found");
+        }
+        Lecture lecture = lectureOptional.get();
+        if (lecture.getCourse() == null) {
+            throw new ConflictException();
+        }
+        if (!authorizationCheckService.isAtLeastEditorInCourse(lecture.getCourse(), null)) {
+            throw new AccessForbiddenException(NOT_ALLOWED);
+        }
```
Comment (b-fein) [inline lecture/src/main/java/de/tum/in/www1/artemis/lecture/web/rest/ExerciseUnitResource.java:74]:
```suggestion
        authorizationCheckService.hasAtLeastRoleInCourseElseThrow(Role.EDITOR, lecture.getCourse(), null));
```
The authorisation checks here and in the other endpoint could be simplified, as no special message is put into the `AccessForbiddenException`.

**Review by b-fein — APPROVED** [review]
Code looks good (one nitpick), works as expected in the manual tests as well. :+1:

File: src/test/java/de/tum/in/www1/artemis/ExerciseUnitIntegrationTest.java [hunk 3]
```diff
@@ -20,6 +20,7 @@
 import de.tum.in.www1.artemis.repository.*;
 import de.tum.in.www1.artemis.util.ModelFactory;
 
+@Deprecated // Moved to Lecture microservice. To be removed
```
Comment (julian-christl) [inline src/test/java/de/tum/in/www1/artemis/ExerciseUnitIntegrationTest.java:23]: Can be removed as indicated in the lecture unit PR

File: lecture/src/main/java/de/tum/in/www1/artemis/lecture/web/rest/ExerciseUnitResource.java (hunk 2 above)
Comment (julian-christl) [reply lecture/src/main/java/de/tum/in/www1/artemis/lecture/web/rest/ExerciseUnitResource.java:74]: This probably can be done in the other Resource classes as well

**Review by julian-christl — APPROVED** [review]
Code looks good

File: lecture/src/main/java/de/tum/in/www1/artemis/lecture/web/rest/ExerciseUnitResource.java (hunk 1 above)
Comment (MerlinMehmed) [reply lecture/src/main/java/de/tum/in/www1/artemis/lecture/web/rest/ExerciseUnitResource.java:85]: Addressed in [#4466](https://github.com/ls1intum/Artemis/pull/4466/commits/568ef4cc8b60e5ab1b64819faf5829cfecea32dd)

**Pull request merged**